qgis/QGIS_Tasks.md

```diff
@@ -0,0 +1,454 @@
+
+```python
+
+class ElevationTileDownloaderDialog(QDialog, Ui_ElevationTileDownloaderDialogBase):
+	...
+    def download_tiles(self) -> None:
+    	...
+        # Create a progress bar in the QGIS message bar
+        self.progress_bar = QProgressBar()
+        self.progress_bar.setMaximum(100)
+        self.progress_bar.setAlignment(Qt.AlignLeft | Qt.AlignVCenter)
+        progress_msg: QgsMessageBarItem = (
+            self.iface.messageBar().createMessage("Download Progress: ")
+        )
+        progress_msg.layout().addWidget(self.progress_bar)
+        self.iface.messageBar().pushWidget(progress_msg, Qgis.Info)
+
+```
+
+Before we show you how we'll update the progress bar, we'll first create our `QgsTask`.
+
+## 3 The `QgsTask`
+
+Generally, there are three ways in which tasks can be created:
+
+- creating it directly from a function,
+- from a processing algorithm,
+- or by extending `QgsTask`
+
+Our task is a little bigger and we like things to be encapsulated, so we chose to extend `QgsTask`. There is a bunch of functions you will need to override/implement when subclassing `QgsTask`:
+- `run()`: this is where the real work will happen in a separate thread, similar to a `QgsProcessingAlgorithm::processAlgorithm`. This method must not call any UI related functions or interact with the main QGIS interface at all or crashes will occur
+- `finished()`: will be called after the `run()` method is finished. It's safe to call UI functions here e.g. to inform the user about success or errors.
+
+### 3.1 Extending `QgsTask`
+We start by creating a new Python module `task.py` inside a new package called `core`. Here, we subclass `QgsTask`:
+```python
+from typing import List, Optional, Set
+from pathlib import Path
+import gzip
+from io import BytesIO
+
+from qgis.gui import QgisInterface
+from qgis.core import (
+    QgsTask,
+    QgsMessageLog,
+    Qgis,
+    QgsRasterLayer,
+    QgsProject,
+    QgsNetworkAccessManager,
+    QgsNetworkReplyContent,
+    QgsGeometry,
+)
+from qgis.utils import iface
+from qgis.PyQt.QtCore import QUrl
+from qgis.PyQt.QtWidgets import QMessageBox
+from qgis.PyQt.QtNetwork import QNetworkRequest
+
+iface: QgisInterface
+
+
+class DownloadTask(QgsTask):
+    def __init__(self, grid: Set[QgsGeometry], output_dir: Path):
+        super().__init__("Tile download task", QgsTask.CanCancel)
+        self.grid = grid
+        self.output_dir = output_dir
+
+        self.exception: Optional[Exception] = None
+        self.tiles: List[Path] = []
+```
+
+The `__init__` is simple: we initialize the parent and assign the grid and the output directory as instance attributes. Note the exception attribute here: since it's not safe to raise exceptions in a background task (this will crash QGIS), we initialize an empty `exception` instance attribute here, where we will store any exception that we catch in `run()`. 
```

Comment:
```suggestion
The task needs the pre-computed grid of elevation tile bboxes and the directory where the user wants to store all those tiles. Also note the exception attribute here: since it's not safe to raise exceptions in a background task (this will crash QGIS), we initialize an empty `exception` instance attribute here, where we will store any exception that we catch in `run()`. 
```